### PR TITLE
[SPARK-10811] [SQL] Eliminates unnecessary byte array copying

### PR DESCRIPTION
**Pull request by liancheng**
When reading Parquet string and binary-backed decimal values, Parquet `Binary.getBytes` always returns a copied byte array, which is unnecessary. Since the underlying implementation of `Binary` values there is guaranteed to be `ByteArraySliceBackedBinary`, and Parquet itself never reuses underlying byte arrays, we can use `Binary.toByteBuffer.array()` to steal the underlying byte arrays without copying them.

This brings performance benefits when scanning Parquet string and binary-backed decimal columns. Note that, this trick doesn't cover binary-backed decimals with precision greater than 18.

My micro-benchmark result is that, this brings a ~15% performance boost for scanning TPC-DS `store_sales` table (scale factor 15).

Another minor optimization done in this PR is that, now we directly construct a Java `BigDecimal` in `Decimal.toJavaBigDecimal` without constructing a Scala `BigDecimal` first. This brings another ~5% performance gain.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/CatalystRowConverter.scala [hunk 1]
```diff
@@ -332,24 +338,30 @@ private[parquet] class CatalystRowConverter(
...
         }
 
-        val bits = 8 * bytes.length
+        val bits = 8 * (end - start)
         unscaled = (unscaled << (64 - bits)) >> (64 - bits)
         Decimal(unscaled, precision, scale)
       } else {
         // Otherwise, resorts to an unscaled `BigInteger` instead.
-        Decimal(new BigDecimal(new BigInteger(bytes), scale), precision, scale)
+        Decimal(new BigDecimal(new BigInteger(value.getBytes), scale), precision, scale)
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/CatalystRowConverter.scala]: For decimals whose precision is greater than 18, we still need to copy the byte array anyway to construct Java `BigInteger` instances.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/CatalystRowConverter.scala [hunk 2]
```diff
@@ -302,7 +302,13 @@ private[parquet] class CatalystRowConverter(
     }
 
     override def addBinary(value: Binary): Unit = {
-      updater.set(UTF8String.fromBytes(value.getBytes))
+      // The underlying `ByteBuffer` implementation is guaranteed to be `HeapByteBuffer`, so here we
+      // are using `Binary.toByteBuffer.array()` to steal the underlying byte array without copying
+      // it.
+      val buffer = value.toByteBuffer
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/CatalystRowConverter.scala]: this still allocates the byte buffer object, doesn't it?

Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/CatalystRowConverter.scala]: Yeah, it allocates a wrapping ByteBuffer. I haven't tried your way yet (using fake `DataOutput` and `Binary.writeTo()` to steal the underlying array). Will give it a shot.

Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/CatalystRowConverter.scala]: I didn't mean to try mine - I think the best way is to change Parquet to give us exactly what we want.

**Comment by liancheng** [comment]
Hm, after some profiling, I found that stealing byte arrays using `ByteBuffer` is actual faster than the `ByteArrayThief` hack I employed in the 2nd commit. This is probably because of virtual function dispatching cost exposed by `OutputStream.write`. Will revert my last commit.

**Comment by liancheng** [comment]
The [last commit] [1] brings another ~5% performance boost. We should probably just use Java `BigDecimal` directly within Catalyst `Decimal`. I guess this would make decimals with large precision (> 18) a little bit faster. But that can be done in a separate PR.

[1]: https://github.com/liancheng/spark/commit/851f91fdcd9a82ef314d4436b768933f8c54875c

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala [hunk 3]
```diff
@@ -145,7 +145,13 @@ final class Decimal extends Ordered[Decimal] with Serializable {
     }
   }
 
-  def toJavaBigDecimal: java.math.BigDecimal = toBigDecimal.underlying()
+  def toJavaBigDecimal: java.math.BigDecimal = {
+    if (decimalVal.ne(null)) {
+      decimalVal.underlying()
+    } else {
+      java.math.BigDecimal.valueOf(longVal, _scale)
```
Comment (liancheng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala]: Originally, we always construct a Scala `BigDecimal` first, and then retrieve the underlying Java `BigDecimal`. Here we just create the Java one directly.

**Comment by davies** [comment]
LGTM

**Comment by cloud-fan** [comment]
LGTM

**Comment by liancheng** [comment]
@davies @cloud-fan Thanks for the review! I'm merging this to master.

**Pull request closed without merging**